src/arithmetic_impls.rs

```diff
@@ -0,0 +1,294 @@
+use crate::{decimal::CalculationResult, ops, Decimal};
+use core::ops::{Add, Div, Mul, Rem, Sub};
+use num_traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedRem, CheckedSub};
+
+// Macros and `Decimal` implementations
+
+macro_rules! impl_checked {
+    ($(#[$($doc:meta)*])* $fun:ident, $impl:ident) => {
+        /// Like the checked version but returns lower or upper bounds instead of `None`.
+        $(#[$($doc)*])*
+        #[inline(always)]
+        #[must_use]
+        pub fn $fun(self, other: Decimal) -> Option<Decimal> {
+            match ops::$impl(&self, &other) {
+                CalculationResult::Ok(result) => Some(result),
+                _ => None,
+            }
+        }
+    };
+}
+
+macro_rules! impl_checked_and_saturating {
+    (
+        $(#[$($checked_doc:meta)*])*
+        $checked_fun:ident,
+        $checked_impl:ident,
+
+        $(#[$($saturating_doc:meta)*])*
+        $saturating_fun:ident
+    ) => {
+        impl_checked!($(#[$($checked_doc)*])* $checked_fun, $checked_impl);
+        impl_saturating!($(#[$($saturating_doc)*])* $saturating_fun, $checked_fun);
+    };
+}
+
+macro_rules! impl_saturating {
+    ($(#[$($doc:meta)*])* $fun:ident, $impl:ident) => {
+        $(#[$($doc)*])*
+        #[inline(always)]
+        #[must_use]
+        pub fn $fun(self, other: Decimal) -> Decimal {
+            if let Some(elem) = self.$impl(other) {
+                elem
+            }
+            else {
+                if self.is_sign_positive() {
+                    Decimal::MAX
+                }
+                else {
+                    Decimal::MIN
+                }
+            }
+        }
+    };
+}
+
+impl Decimal {
+    impl_checked_and_saturating!(
+        /// Checked addition. Computes `self + other`, returning `None` if overflow occurred.
+        ///
+        /// ```rust
+        /// use rust_decimal::Decimal;
+        ///
+        /// assert_eq!(Decimal::MAX.checked_add(Decimal::MAX), None);
+        /// assert_eq!(Decimal::MIN.checked_add(Decimal::MIN), None);
+        /// ```
+        checked_add,
+        add_impl,
+        /// ```rust
+        /// use rust_decimal::Decimal;
+        ///
+        /// assert_eq!(Decimal::MAX.saturating_add(Decimal::ONE), Decimal::MAX);
+        /// assert_eq!(Decimal::MIN.saturating_add(Decimal::MIN), Decimal::MIN);
+        /// ```
+        saturating_add
```

Comment:
The `saturating` prefix is the prefix used by the standard library, i.e., probably the canonical naming